zng/docs/zson.md

```diff
@@ -0,0 +1,629 @@
+record type implied by each of the three variations of type of the `value` field.
+
+## The ZSON Data Model
+
+ZSON data is defined as an ordered sequence of one or more typed data values
+separated at any point or not at all by an end-of-sequence marker `.`.
+Each sequence implies a type context as described below and the `.` marker causes
+a new type context to be established for subsequent elements of the sequence.
+
+Each value's type is either a "primitive type", a "complex type", the "type type",
+or the "null type".
+
+### Primitive Types
+
+Primitive types include signed and unsigned integers, IEEE floating point of
+several widths, IEEE decimal,
+string, bstring, byte sequence, boolean, IP address, and IP network.
+
+> Note: The `bstring` type is an unusual type representing a hybrid type
+> mixing a UTF-8 string with embedded binary data.  This type is
+> useful in systems that, for instance, pull data off the network
+> while expecting a string, but there can be embedded binary data due to
+> bugs, malicious attacks, etc.  It is up to the receiver to determine
+> with out-of-band information or inference whether the data is ultimately
+> arbitrary binary data or a valid UTF-8 string.
+
+### Complex Types
+
+Complex types are composed of primitive types and/or other complex types
+and include
+* _record_ - an ordered collection of zero or more named values called fields,
+* _array_ - an ordered sequence of zero or more values called elements,
+* _set_ - a set of zero or more unique values called elements,
+* _union_ - a type representing values whose type is any of a specified collection of two or more types,
+* _enum_ - a type representing values taken from a specified collection of one or more uniformly
+typed values, which can be either primitive or complex, and
+* _map_ - a collection of zero or more key/value pairs where the keys are of a
+uniform type called the key type and the values are of a uniform type called
+the value type.
+
+### The Type Type
+
+ZSON also includes first-class types, where a value can be of type `type`.
+The syntax for type values is given below.
+
+### The Null Type
+
+The _null_ type is a primitive type representing only a `null` value.  A `null`
+value can have any type.
+
+Any value in ZSON can take on a null representation.  It is up to an
+implementation to decide how external data structures map into and
+out of values with nulls.  Typically, a null value means either the
+zero value or in the case of record fields, it means the field is optional
+and the value is not present, though these semantics are not explicitly
+defined by ZSON.
+
+### Type Definitions and Contexts
+
+Type definitions embedded in the data sequence associate a name with a type
+so that later values can refer to their type by name instead of explicitly
+enumerating the type of every element.  A collection of mappings from name to
+type is called a "type context."  As bindings are read from a sequence, they
```

Comment:
I think this reads better with consistent use of "binding" terminology.
```suggestion
Type definitions embedded in the data sequence bind a name to a type
so that later values can refer to their type by name instead of explicitly
enumerating the type of every element.  A collection of bindings from name to
type is called a "type context."  As bindings are read from a sequence, they
```